### PR TITLE
Control plane Discoverer

### LINKED ISSUE #241: Make discoverer easier to test
At the moment `Discoverer` interface https://github.com/newrelic/nri-kubernetes/blob/0a78aa518374cc8d250122d26268ba5e1858a754/src/client/discovery.go#L8-L11 returns an `HTTPClient` interface https://github.com/newrelic/nri-kubernetes/blob/0a78aa518374cc8d250122d26268ba5e1858a754/src/client/discovery.go#L20-L24 which makes it difficult to test, as it is obfuscated where the request will be sent.

Most likely we should decompose `Discoverer` interface to make it have an endpoint discovery part and client configuration part, so we can only test discovered endpoints without caring about client configuration and performing the requests.

### PR DESCRIPTION
**Pull request by gsanchezgavier**
Moves pod discover logic to Discoverer, and add some corner cases testing.
Fixes #241

### CODE REVIEW

File: src/controlplane/discoverer/discoverer_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+)
+
+func Test_Discoverer_does_not_fail(t *testing.T) {
+	t.Parallel()
+
+	namespace := "testNamespace"
+	nodeName := "testNode"
+	selector := "foo=bar"
+
+	testCases := []struct {
```
Comment (paologallinaharbur) [inline src/controlplane/discoverer/discoverer_test.go:27]: I might missed it, but I would write a tests and add a comments saying what happens if the list of selector provided is empty

**Review by paologallinaharbur — APPROVED** [review]
Only a doubt regarding a behaviour I am not sure we want to allow permit, everything else looks good!

File: src/controlplane/discoverer/discoverer.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/internal/discovery"
+)
+
+// PodDiscoverer is used to discover control plane components.
+type PodDiscoverer interface {
+	// Discover returns looks for a pod mathching the selector, namespaces and
```
Comment (roobre) [inline src/controlplane/discoverer/discoverer.go:16]: returns or looks?

File: src/controlplane/discoverer/discoverer.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	labelsSet, err := labels.ConvertSelectorToLabelsMap(ad.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid selector %q: %w", ad.Selector, err)
+	}
+
+	selector := labels.SelectorFromSet(labelsSet)
+
+	pods, err := podLister.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("fail to list pods for selector: %v", labelsSet)
```
Comment (roobre) [inline src/controlplane/discoverer/discoverer.go:78]:
```suggestion
		return nil, fmt.Errorf("listing pods with selector %q: %w", labelsSet, err)
```

File: src/controlplane/discoverer/discoverer.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	for _, pod := range pods {
+		if ad.MatchNode && pod.Spec.NodeName != c.Config.NodeName {
+			c.logger.Debugf("Discarding pod: %s running outside the node", pod.Name)
+			continue
+		}
+		// first pod matching all conditions is returned.
+		return pod, nil
+	}
+
+	return nil, nil
```
Comment (roobre) [inline src/controlplane/discoverer/discoverer.go:92]: Should we return a non-nil error here? Like a `ErrNoPodsFound`.

**Review by roobre — APPROVED** [review]
Looking great so far! Just a couple nits

File: src/controlplane/discoverer/discoverer.go (hunk 4 above)
Comment (gsanchezgavier) [reply src/controlplane/discoverer/discoverer.go:92]: yes it make sense! at least just to have this custom error 👍

File: src/controlplane/discoverer/discoverer_test.go (hunk 1 above)
Comment (gsanchezgavier) [reply src/controlplane/discoverer/discoverer_test.go:27]: just added on "when_empty_selector", thanks

**Pull request merged**